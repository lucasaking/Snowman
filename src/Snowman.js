import React, { useState } from "react";

import "./Snowman.css";
import img0 from "./0.png";
import img1 from "./1.png";
import img2 from "./2.png";
import img3 from "./3.png";
import img4 from "./4.png";
import img5 from "./5.png";
import img6 from "./6.png";


/** Snowman game: plays hangman-style game with a melting snowman.
 *
 * Props:
 * - maxWrong: how many wrong moves is a player allowed?
 * - images: array of images for wrong guess
 * - words: array of words to pick answer from
 *
 * State:
 * - nWrong: # wrong guesses so far
 * - guessed: set of guessed letters (good and bad) so far
 * - answer: selected secret word*
 */

function Snowman({
  maxWrong = 6,
  images = [img0, img1, img2, img3, img4, img5, img6],
  words = ["apple"] }) {

  /** by default, allow 6 guesses and use provided gallows images. */

  // Add new state for randomWord 
  const [nWrong, setNWrong] = useState(0); //Add the number of wrong guesses
  const [guessed, setGuessed] = useState(new Set());
  const [answer, setAnswer] = useState((words)[0]); //random word goes here?

  /** guessedWord: show current-state of word:
   if guessed letters are {a,p,e}, show "app_e" for "apple"
   */
  function guessedWord() {
    return answer
      .split("")
      .map(ltr => (guessed.has(ltr) ? ltr : "_"));
  }

  /** handleGuess: handle a guessed letter:
   - add to guessed letters
   - if not in answer, increase number-wrong guesses
   */
  function handleGuess(evt) {
    let ltr = evt.target.value;

    setGuessed(g => {
      console.log("THISS -->", g);
      const newGuessed = new Set(g); // Creates a set from the ltrs chosen
      newGuessed.add(ltr); // Adds ltr to the new set 
      return newGuessed; // return set
    });

    setNWrong(n => n + (answer.includes(ltr) ? 0 : 1)); // Change 'setWrong' state-- if it includes ltr dont change image
  }

  /** generateButtons: return array of letter buttons to render */
  function generateButtons() {
    return "abcdefghijklmnopqrstuvwxyz".split("").map(ltr => ( // 
      <button
        key={ltr} //unique value 
        value={ltr} //
        onClick={handleGuess} //Run handleGuess on click, which will return a new set of updated ltrs
        disabled={guessed.has(ltr)} //ltr has already been chosen, prevents clicking again
      >
        {ltr}
      </button> //{ltr} === ^display the letter on button 
    ));
  }

  if (nWrong === maxWrong) {
    return (<div>
      <h3 className="lose">You lose!</h3><h5>{words[0]}</h5>
    </div>)
  }
  /** render: render game */
  return (
    <div className="Snowman">
      <img src={(images)[nWrong]} alt={nWrong} />
      <p>Number of Wrong Guesses: {nWrong}</p>
      <p className="Snowman-word">{guessedWord()}</p>
      <p>{generateButtons()}</p>
    </div>
  ); //^ Should these functions have ()?
}

export default Snowman;
