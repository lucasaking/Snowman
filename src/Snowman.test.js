import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Snowman from "./Snowman";

// it("works when you click on the right arrow", function () {
//   const { queryByTestId, queryByAltText } = render(<Carousel />);

//   // expect the first image to show, but not the second
//   expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
//   expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();

//   // move forward in the carousel
//   const rightArrow = queryByTestId("right-arrow");
//   fireEvent.click(rightArrow);

//   // expect the second image to show, but not the first
//   expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
//   expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();
// });

const MAX_WRONG = 6;

it("renders without crashing", function () {
  // this is a low-value test, but better than nothing
  render(<Snowman
    maxWrong={MAX_WRONG}
    images="6.png"
    words={"apple"}
  />);
});

it("matches snapshot", function () {
  const { container } = render(<Snowman
    maxWrong={6}
    images="6.png"
    words={"apple"}
  />);
  expect(container).toMatchSnapshot();
});

it("renders crashes after 6 guesses", function () {
  // the app continues to show images after maxGuesses
  
  const { container, getByText } = render(<Snowman maxWrong={1}/>);
  fireEvent.click(getByText("r"));
  getByText("You lose!");
  getByText("apple");

  expect(container).toMatchSnapshot();
});
