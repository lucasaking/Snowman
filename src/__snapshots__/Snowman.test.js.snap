// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="Snowman"
  >
    <img
      alt="0"
      src="6"
    />
    <p>
      Number of Wrong Guesses: 
      0
    </p>
    <p
      class="Snowman-word"
    >
      _
    </p>
    <p>
      <button
        value="a"
      >
        a
      </button>
      <button
        value="b"
      >
        b
      </button>
      <button
        value="c"
      >
        c
      </button>
      <button
        value="d"
      >
        d
      </button>
      <button
        value="e"
      >
        e
      </button>
      <button
        value="f"
      >
        f
      </button>
      <button
        value="g"
      >
        g
      </button>
      <button
        value="h"
      >
        h
      </button>
      <button
        value="i"
      >
        i
      </button>
      <button
        value="j"
      >
        j
      </button>
      <button
        value="k"
      >
        k
      </button>
      <button
        value="l"
      >
        l
      </button>
      <button
        value="m"
      >
        m
      </button>
      <button
        value="n"
      >
        n
      </button>
      <button
        value="o"
      >
        o
      </button>
      <button
        value="p"
      >
        p
      </button>
      <button
        value="q"
      >
        q
      </button>
      <button
        value="r"
      >
        r
      </button>
      <button
        value="s"
      >
        s
      </button>
      <button
        value="t"
      >
        t
      </button>
      <button
        value="u"
      >
        u
      </button>
      <button
        value="v"
      >
        v
      </button>
      <button
        value="w"
      >
        w
      </button>
      <button
        value="x"
      >
        x
      </button>
      <button
        value="y"
      >
        y
      </button>
      <button
        value="z"
      >
        z
      </button>
    </p>
  </div>
</div>
`;

exports[`renders crashes after 6 guesses 1`] = `
<div>
  <div>
    <h3
      class="lose"
    >
      You lose!
    </h3>
    <h5>
      apple
    </h5>
  </div>
</div>
`;
